Use protocol buffers for internal StorageAPI wire encoding

vekterli opened the pull request:
@geirst please review

This is an implementation of all StorageAPI requests and responses in Protocol Buffers, replacing the existing hand-rolled binary codecs.

This helps remove a lot of tech debt and lets us be vastly more agile with protocol changes going forward. Old binary protocol implementations will be deprecated after this and should be removed on Vespa 8.

Must not be merged until we have decided what exact (near future) version to tag the protocol with.
Also pending any internal reviews of protobuf schema.

vekterli committed:
Convert protocol serialization tests from CppUnit to GTest

vekterli committed:
Make serialization tests version-parameterizable

vekterli committed:
Add start of protobuf serialization protocol implementation

vekterli committed:
Add support for DeleteBucket, Merge and GetBucketDiff

vekterli committed:
Add support for ApplyBucketDiff and RequestBucketInfo

vekterli committed:
Implement NotifyBucketChange, SplitBucket and JoinBuckets

vekterli committed:
Implement SetBucketState, Create/DestroyVisitor and RemoveLocation

vekterli committed:
Don't derive v7 serialization from a prior versioned protocol

Make `ProtocolSerialization` a proper interface and introduce an
intermediate `LegacyProtocolSerialization` class for the legacy
versioned protocol implementations.

vekterli committed:
Break up protobufs into multiple files to avoid single, massive generated .h/.cc

vekterli committed:
Faster protobuf serialization without `std::string` indirection

vekterli committed:
Remove old 5.0 code path that generated data no one used

vekterli committed:
Refactor utility codec functions to build on top of each other

Also move protobuf includes out to a dedicated header file that hides
the various cozy GCC warning suppressions we need for the `protoc`-
generated code.

vekterli committed:
Factor out shared protobuf code

vekterli committed:
Extend testing of encoded/decoded fields

vekterli committed:
Misc cleanup and refactoring

vekterli committed:
Add missing header pragma directive

vekterli committed:
Explicitly test response without bucket ID remapping

geirst commented on storageapi/src/tests/mbusprot/storageprotocoltest.cpp:
```diff
@@ -32,183 +36,108 @@ using document::test::makeBucketSpace;
     vespalib::Version _version5_1{5, 1, 0};
     vespalib::Version _version5_2{5, 93, 30};
     vespalib::Version _version6_0{6, 240, 0};
+    vespalib::Version _version7_0{7, 0, 0}; // FIXME
```
All these version variables can be removed?

vekterli replied:
Done

geirst commented on storageapi/src/vespa/storageapi/mbusprot/protobuf/common.proto:
```diff
@@ -0,0 +1,93 @@
+// Note: we use a *Request/*Response naming convention rather than *Command/*Reply,
+// as the former is the gRPC convention and that's where we intend to move.
+
+// Next tag to use: 2
```
Re discussion: Remove next tag comments in favor of `reserved` if / when removing a field.

vekterli replied:
Done

geirst commented on storageapi/src/vespa/storageapi/mbusprot/protobuf/common.proto:
```diff
@@ -0,0 +1,93 @@
+}
+
+// Next tag to use: 9
+message BucketInfoV1 {
```
Re discussion:
- Remove BucketInfoV2 and BucketInfo
- Rename BucketInfoV1 to BucketInfo
- Rename checksum to legacy_checksum
- Create message for new checksum and put into BucketInfo

vekterli replied:
Done, although I haven't added a new message for the V2 checksum since it's not yet fully spec'd. We can now finally take that kind of liberty since adding message fields is painless! 🙂🦄

geirst reviewed (approved):
👍 Overall looks, some comments to consider

vekterli committed:
Address code review feedback

- don't state next tag in comments, we'll use `reserved` instead
- restructure bucket info to version the checksum and not the info itself
- add some comments

vekterli commented on storageapi/src/vespa/storageapi/mbusprot/legacyprotocolserialization.h:
```diff
@@ -0,0 +1,26 @@
+
+namespace storage::mbusprot {
+
+class LegacyProtocolSerialization : public ProtocolSerialization {
```
Done

vekterli commented:
Explicit version still needs to be committed prior to merge (as well as no merge stop, of course)

geirst reviewed (approved):
👍 🏅

vekterli committed:
Tag protocol switchover version as 7.40.5
